docs/design/operator-bundle.md

```diff
@@ -0,0 +1,199 @@
+* The label `operators.operatorframework.io.bundle.metadata.v1` reflects the path in the image to the directory that contains metadata files about the bundle.
+* The `manifests.v1` and `metadata.v1` labels imply the bundle type:
+    * The value `manifests.v1` implies that this bundle contains operator manifests.
+    * The value `metadata.v1` implies that this bundle has operator metadata.
+* The label `operators.operatorframework.io.bundle.package.v1` reflects the package name of the bundle.
+* The label `operators.operatorframework.io.bundle.channels.v1` reflects the list of channels the bundle is subscribing to when added into an operator registry
+* The label `operators.operatorframework.io.bundle.channel.default.v1` reflects the default channel an operator should be subscribed to when installed from a registry
+
+The labels will also be put inside a YAML file, as shown below.
+
+*annotations.yaml*
+```yaml
+annotations:
+  operators.operatorframework.io.bundle.mediatype.v1: "registry+v1"
+  operators.operatorframework.io.bundle.manifests.v1: "manifests/"
+  operators.operatorframework.io.bundle.metadata.v1: "metadata/"
+  operators.operatorframework.io.bundle.package.v1: "test-operator"
+  operators.operatorframework.io.bundle.channels.v1: "beta,stable"
+  operators.operatorframework.io.bundle.channel.default.v1: "stable"
+```
+
+*Notes:*
+* In case of a mismatch, the `annotations.yaml` file is authoritative because the on-cluster operator-registry that relies on these annotations has access to the yaml file only.
+* The potential use case for the `LABELS` is - an external off-cluster tool can inspect the image to check the type of a given bundle image without downloading the content.
+
+This example uses [Operator Registry Manifests](https://github.com/operator-framework/operator-registry#manifest-format) format to build an operator bundle image. The source directory of an operator registry bundle has the following layout.
+```
+$ tree test
+test
+├── testbackup.crd.yaml
+├── testcluster.crd.yaml
+├── testoperator.v0.1.0.clusterserviceversion.yaml
+├── testrestore.crd.yaml
+└── metadata
+    └── annotations.yaml
+```
+
+### Bundle Dockerfile
+
+This is an example of a `Dockerfile` for operator bundle:
+```
+FROM scratch
+
+# We are pushing an operator-registry bundle
+# that has both metadata and manifests.
+LABEL operators.operatorframework.io.bundle.mediatype.v1=registry+v1
+LABEL operators.operatorframework.io.bundle.manifests.v1=manifests/
+LABEL operators.operatorframework.io.bundle.metadata.v1=metadata/
+LABEL operators.operatorframework.io.bundle.package.v1=test-operator
+LABEL operators.operatorframework.io.bundle.channels.v1=beta,stable
+LABEL operators.operatorframework.io.bundle.channel.default.v1=stable
+
+ADD test/*.yaml /manifests
+ADD test/metadata/annotations.yaml /metadata/annotations.yaml
```

Comment:
if above comment holds, it's more convenient if we don't have the manifests and metadata in different layers (which is what the separate ADD commands will do afaik). So we could change the above to

```
ADD test /
```

Reply:
Give this is a CLI tool that is controlled by us, we want a strict build process that only adds yaml manifests + metadata (`annotations.yaml`) that we either generate or validate. We want to avoid adding everything in the directory which may include unwanted garbage such as non-consumable files or binaries by OLM. One of the side goals of this enhancement is to keep the image to be as small as possible. If adding everything in the directory, we potentially add a ton of stuff that users may leave there and inflate the image size.

Reply:
I agree with building the manifests and metadata to avoid unwanted garbage. Can we add these folders to the same layer? Is there a reason we want separate layers for each folder?

Reply:
Give the ADD can't be chained, not sure how I can add both in the same layer. They are going in two separate directories as well.

Reply:
That's a good point about chaining not being supported. Ok I guess we have to do it this way.

Reply:
Really should be using COPY here instead, but that can be changed later if anybody else agrees. Is there anything aside from efficiency that is a concern here for how many layers are used?